design-proposals/klip-35-dynamic-processing-log-levels.md

```diff
@@ -0,0 +1,210 @@
+
+For important background on the processing log, please read https://docs.ksqldb.io/en/latest/developer-guide/test-and-debug/processing-log/
+
+## In-Scope
+
+- API for getting/setting the global log level
+- API for getting/setting query/node-level log levels
+- Kafka-based stateful implementation that distributes level settings to the cluster and stores them so they are recovered on a restart.
+
+## Out-Of-Scope
+
+- Thoroughly populating the query engine with log messages to address the problems we’ve already seen. This is super important, but out of scope for this particular KLIP.
+
+## Value/Return
+
+A step toward improving UX when users observe unexpected behavior from running queries.
+
+## Public APIs
+
+### REST API
+
+### PUT /processing_logger/config
+
+Sets the root processing log configuration (currently just supports setting the log level).  This will set the level for all loggers that do not already have a level set.
+
+Request Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+````
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/config
+
+Gets the current root processing log level.
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level of the root logger”}
+    }
+}
+```
+
+### DELETE /processing_logger/config
+
+Deletes customized config for the root logger (reverts to default config as specified in the log4j config file).
+
+### PUT /processing_logger/[processing logger name]/config
+
+Sets the processing log configuration for the logger with the given processing logger name (currently just supports setting the log level). This will set the level for all child loggers (loggers that share a common name prefix) that do not already have a level set using a longer prefix.
+
+Request Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+```
+
+Response Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/[processing logger name]/level
+
+Gets the processing log configuration for the logger with the given processing logger name.
+
+Response Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level of the specified logger”}
+    }
+}
+```
+
+### DELETE /processing_logger/[processing logger name]/level
+
+Deletes customized config for the logger with the given processing logger name.
+
+### GET /processing_logger/[processing logger name?]
```

Comment:
This is a bit confusing in light of `GET /processing_logger/config` which returns just the config. Maybe we can add the root config into the response of `GET /processing_logger/` too, in order to avoid confusion?

Reply:
yeah that was the idea. `GET /processing_logger/` would list all loggers. `GET /processing logger/[processing logger name]` would list all loggers under a given name.

Reply:
> `GET /processing_logger/` would list all loggers.

Does "all loggers" include the root processing log level? (Do we actually have a logger for that, or is that separate?) As in, if a user has only adjusted the root log level but hasn't configured any per-logger overrides, will that show up in `GET /processing_logger/` or only in `GET /processing_logger/config`? My confusion was that `GET /processing_logger/` seems like it should include the information in `GET /processing_logger/config`.

Reply:
Yep - it would include the root processing log as well.

Reply:
So that means `PUT /processing_logger/config` is equivalent to `PUT /processing_logger/root/config`, and similarly for GET and DELETE as well?

Reply:
I think you're getting at some legit problems with this API design. For example, what if a processing log node has "config" in it. I'll rework it or switch to @agavra 's cool idea below